File: _curso/6_ddd_modelagem_tatica_e_patterns/src/domain/checkout/service/order.service.ts
import { v4 as uuid } from "uuid";
import Order from "../../../entity/aggregate/order";
import OrderItem from "../../../entity/aggregate/order_item";
import Customer from "../../../entity/customer";

export default class OrderService {
    static placeOrder(customer: Customer, items: OrderItem[]): Order {
        if (items.length === 0) {
            throw new Error("Order must have at least one item");
        }
        const order = new Order(uuid(), customer.id, items);
        customer.addRewardPoints(order.total() / 2);
        return order;
    }

    static total(orders: Order[]): number {
        return orders.reduce((acc, order) => acc + order.total(), 0);
    }
}

File: 6_ddd_modelagem_tatica_e_patterns/python/src/domain/checkout/service/order_service.py
import uuid
from functools import reduce
from typing import List
from domain.checkout.entity.order import Order
from domain.checkout.entity.order_item import OrderItem
from domain.customer.entity.customer import Customer

class OrderService:
    @staticmethod
    def place_order(customer: Customer, items: List[OrderItem]) -> Order:
        if not items:
            raise Exception("Order must have at least one item")

        order = Order(str(uuid.uuid4()), customer.id, items)
        customer.add_reward_points(order.total / 2)
        return order

    @staticmethod
    def total(orders: List[Order]) -> int:
        return reduce(lambda total, order: total + order.total, orders, 0)